### ISSUE TITLE
bbc generates wrong code for BLAS DROTGM

### ISSUE THREAD

**Issue by oroppas** [flang-compiler/f18-llvm-project#819]
`bbc` seems to generate wrong mlir. Tested on one of the BLAS routine, `DROTGM` (the same goes to `SROTGM`):

http://www.netlib.org/lapack/explore-html/de/da4/group__double__blas__level1_ga13e351a3dfafa2cd8dc5302dcf53f69a.html#ga13e351a3dfafa2cd8dc5302dcf53f69a

The following test code checks if `DROTGM` can reduce the vector `(x, y)^T` into `(r, 0)^T` by computing  Givens rotation `H`
in the form

```
  | x |   | r |
H |   | = |   |
  | y |   | 0 |
```

```fortran
      PROGRAM TEST_DROT

      IMPLICIT NONE

      DOUBLE PRECISION ZERO, ONE
      DATA ZERO, ONE /0.0D+0, 1.0D+0/
      DOUBLE PRECISION  X1, Y1
      DOUBLE PRECISION  DD1, DD2
      DOUBLE PRECISION  DX1, DY1
      DOUBLE PRECISION  DPARAM(5)

      DD1 = 1.0D+0
      DD2 = 1.0D+0
      DX1 = 0.9134D+0
      DY1 = 0.6324D+0
      X1 = DX1
      Y1 = DY1

      PRINT *, 'INPUT'
      PRINT *, '-----'
      PRINT *, 'DD1 = ', DD1
      PRINT *, 'DD2 = ', DD2
      PRINT *, 'DX1 = ', DX1
      PRINT *, 'DY1 = ', DY1
      PRINT *

      CALL DROTMG(DD1, DD2, DX1, DY1, DPARAM)

      PRINT *, 'OUTPUT'
      PRINT *, '------'
      PRINT *, 'DD1 = ', DD1
      PRINT *, 'DD2 = ', DD2
      PRINT *, 'DX1 = ', DX1
      PRINT *, 'DY1 = ', DY1
      PRINT *
      PRINT *, 'DFLAG = ', DPARAM(1)
      PRINT *

      IF ( DPARAM(1) .EQ. -ONE ) THEN
        PRINT *, '|DH11 DH12| ', DPARAM(2), DPARAM(4)
        PRINT *, '|DH21 DH22| ', DPARAM(3), DPARAM(5)
        PRINT *
        PRINT *, '|r| ', DPARAM(2) * X1 + DPARAM(4) * Y1
        PRINT *, '|0| ', DPARAM(3) * X1 + DPARAM(5) * Y1
      ELSE IF ( DPARAM(1) .EQ. ZERO ) THEN
        PRINT *, '|DH11 DH12| ', 1.0D+0, DPARAM(4)
        PRINT *, '|DH21 DH22| ', DPARAM(3), 1.0D+0
        PRINT *
        PRINT *, '|r| ', 1.0D+0    * X1 + DPARAM(4) * Y1
        PRINT *, '|0| ', DPARAM(3) * X1 + 1.0D+0    * Y1
      ELSE IF ( DPARAM(1) .EQ. ONE ) THEN
        PRINT *, '|DH11 DH12| ', DPARAM(2), 1.0D+0
        PRINT *, '|DH21 DH22| ', -1.0D+0, DPARAM(5)
        PRINT *
        PRINT *, '|r| ', DPARAM(2) * X1 + 1.0D+0    * Y1
        PRINT *, '|0| ', -1.0D+0   * X1 + DPARAM(5) * Y1
      ELSE
        PRINT *, '|DH11 DH12| ', 1.0D+0, 0.0D+0
        PRINT *, '|DH21 DH22| ', 0.0D+0, 1.0D+0
        PRINT *
        PRINT *, '|r| ', 1.0D+0 * X1 + 0.0D+0 * Y1
        PRINT *, '|0| ', 0.0D+0 * X1 + 1.0D+0 * Y1
      END IF

      END
```

`gfortran` can produce the expected result

```
 INPUT
 -----
 DD1 =    1.0000000000000000
 DD2 =    1.0000000000000000
 DX1 =   0.91339999999999999
 DY1 =   0.63239999999999996
 OUTPUT
 ------
 DD1 =   0.67596802837255554
 DD2 =   0.67596802837255554
 DX1 =    1.3512473396102473
 DY1 =   0.63239999999999996
 DFLAG =    0.0000000000000000
 |DH11 DH12|    1.0000000000000000       0.69235822202758923
 |DH21 DH22|  -0.69235822202758923        1.0000000000000000
 |r|    1.3512473396102473
 |0|    0.0000000000000000
```

whereas `flang` generate wrong result (note that the second entry of `(r 0)^T` is non-zero)

```
 INPUT
 -----
 DD1 =  1.
 DD2 =  1.
 DX1 =  .9134
 DY1 =  .6324
 OUTPUT
 ------
 DD1 =  .6759680283725555
 DD2 =  .6759680283725555
 DX1 =  1.3512473396102473
 DY1 =  .6324
 DFLAG =  -1.
 |DH11 DH12|  1. 1.
 |DH21 DH22|  -2.44140625E-04 1.
 |r|  1.5457999999999998
 |0|  .632177001953125
```

Offending lines in `DROTGM` are

```fortran
193             DO WHILE ((dd1.LE.rgamsq) .OR. (dd1.GE.gamsq))
```

and

```fortran
218             DO WHILE ( (dabs(dd2).LE.rgamsq) .OR. (dabs(dd2).GE.gamsq) )
```

Inserting print statement before `DO WHILE` changes the behavior of the generated code
and produce the correct result.

Snippet of diff of generated `mlir` with and without single print statement before `DO WHILE`

![image](https://user-images.githubusercontent.com/885380/120591549-4bfdab00-c477-11eb-8bdb-0af8c9fb9d16.png)

**Comment by jeanPerier** [comment]
Thanks for the report, that's indeed a bug in the IR we are generating. Your analysis is correct, there is an issue with the IF/DO WHILE succession.
 
```fortran
subroutine bad(b1, b2)
  logical :: b1, b2
  if (b1) then
  ! No problem if statement here
  ! call something()
  do while (b2)
    call foo()
   end do
  end if
end subroutine
```

`bbc -emit-fir` output is:

```
func @_QPbad(%arg0: !fir.ref<!fir.logical<4>>, %arg1: !fir.ref<!fir.logical<4>>) {
  %0 = fir.load %arg0 : !fir.ref<!fir.logical<4>>
  %1 = fir.convert %0 : (!fir.logical<4>) -> i1
  cond_br %1, ^bb1, ^bb4
^bb1:  // pred: ^bb0
  br ^bb3
^bb2:  // pred: ^bb3
  %2 = fir.load %arg1 : !fir.ref<!fir.logical<4>>
  %3 = fir.convert %2 : (!fir.logical<4>) -> i1
  cond_br %3, ^bb3, ^bb4
^bb3:  // 2 preds: ^bb1, ^bb2
  fir.call @_QPfoo() : () -> ()
  br ^bb2
^bb4:  // 2 preds: ^bb0, ^bb2
  return
}
```



It looks like `^bb1` should branch to `^bb2`, not `^bb3`.

@vdonaldson , it looks like an unstructured if construct generation issue.

**Comment by vdonaldson** [comment]
If a DO WHILE loop is a branch target, and therefore needs a preheader block, the preheader block should contain (only) a branch to the loop header block.

**Comment by jeanPerier** [comment]
Fixed by #824.
